src/backend/executor/execMain.c

```diff
@@ -972,6 +973,61 @@ standard_ExecutorRun(QueryDesc *queryDesc,
 		}
 		if ((exec_identity == GP_IGNORE || exec_identity == GP_ROOT_SLICE) && operation != CMD_SELECT)
 			es_processed = mppExecutorWait(queryDesc);
+
+		/*
+		 * Update view info if possible.
+		 * Use the operation and result relation to indentify if
+		 * a table's data is changed.
+		 * This is a proper place for all tables of different AMs.
+		 * We handle INSERT/UPDATE/DELETE here as other operations should
+		 * be handled in utility commands.
+		 *
+		 * Use es_processed to indentify the actual rows we modified
+		 * to avoid case writablte query may not
+		 * change data when success, ex:
+		 *  insert into t1 select * from t2;
+		 * When t2 has zero rows, don't need to update view status.
+		 *
+		 * NB: This can't handle well in utility mode, should REFRESH by user
+		 * after that.
+		 */
+		if ((GP_ROLE_DISPATCH == Gp_role && es_processed > 0) ||
+			(IS_SINGLENODE() && (operation != CMD_SELECT) && estate->es_processed > 0))
```

Comment:
We also need to catch for  'BEFORE ROW INSERT or INSTEAD OF ROW INSERT' triggers?

Reply:
NSTEAD OF ROW INSERT in only for  updatable view which is not well supported in CBDB, and we only support one base table to maintain view datastaus.
So it could be bypassed in CBDB or SingleNode mode.

BEFORE ROW INSERT
 * for triggers on segments, CBDB could not exec a non-SELECT function: `function cannot execute on a QE slice because it issues a non-SELECT statement  `
 * for triggers try to modify entry tables, most of them are catalog which is not allowed in production.
 
In SingleNode mode, all tables include user tables are on QD.
Need to verify BEFORE ROW INSERT triggers and etc.

Reply:
Verify that it could work with triggers  in SingleNode mode:
tri_t1 has trigers insert into tri_t2 and update tri_t3.
An insertion into t1 will update tri_t2, tri_t3 relative materialized views data status well.

```sql
create table tri_t1(a int, b int);
create table tri_t2(a int, b int);
create table tri_t3(a int, b int);
create materialized view tri_mv1 as select * from tri_t1;
create materialized view tri_mv2 as select * from tri_t2;
insert into tri_t3 values (1, 2);
create materialized view tri_mv3 as select * from tri_t3;
create function trigger_insert_tri_t2()
returns trigger AS
$$
begin
  execute 'insert into tri_t2 values(1, 1)';
  execute 'update tri_t3 set b = 10 where a = 1;';
  return NEW;
end;
$$
language plpgsql;
create trigger trigger_insert_tri_t2 before insert ON tri_t1
  for each row execute procedure trigger_insert_tri_t2();
select mvname, datastatus from gp_matview_aux where mvname in ('tri_mv1', 'tri_mv2', 'tri_mv3');
 mvname  | datastatus 
---------+------------
 tri_mv1 | u
 tri_mv2 | u
 tri_mv3 | u
(3 rows)

select * from tri_t1;
 a | b 
---+---
(0 rows)

select * from tri_t2;
 a | b 
---+---
(0 rows)

select * from tri_t3;
 a | b 
---+---
 1 | 2
(1 row)

insert into tri_t1 values(1, 2);
select * from tri_t1;
 a | b 
---+---
 1 | 2
(1 row)

-- should also insert data
select * from tri_t2;
 a | b 
---+---
 1 | 1
(1 row)

-- shoud be updated
select * from tri_t3;
 a | b  
---+----
 1 | 10
(1 row)

-- check mv status
select mvname, datastatus from gp_matview_aux where mvname in ('tri_mv1', 'tri_mv2', 'tri_mv3');
 mvname  | datastatus 
---------+------------
 tri_mv2 | i
 tri_mv3 | e
 tri_mv1 | i
(3 rows)
```
i -> insert only
e -> expired